pkg/dns/dnscache.go

```diff
@@ -0,0 +1,438 @@
+				Version:        firewallv1.IPv6,
+			},
+		)
+	}
+
+	return nil
+}
+
+// Update DNS cache.
+// It expects that there was only one question to DNS(majority of cases).
+// So it picks first qname and skips all others(if there is).
+func (c *DNSCache) Update(lookupTime time.Time, msg *dnsgo.Msg) error {
+	qname := strings.ToLower(msg.Question[0].Name)
+
+	ipv4 := []net.IP{}
+	ipv6 := []net.IP{}
+	minIPv4TTL := uint32(math.MaxUint32)
+	minIPv6TTL := uint32(math.MaxUint32)
+
+	for _, ans := range msg.Answer {
+		if strings.ToLower(ans.Header().Name) != qname {
+			continue
+		}
+
+		switch rr := ans.(type) {
+		case *dnsgo.A:
+			ipv4 = append(ipv4, rr.A)
+			if minIPv4TTL > rr.Hdr.Ttl {
+				minIPv4TTL = rr.Hdr.Ttl
+			}
+		case *dnsgo.AAAA:
+			ipv6 = append(ipv6, rr.AAAA)
+			if minIPv6TTL > rr.Hdr.Ttl {
+				minIPv6TTL = rr.Hdr.Ttl
+			}
+		}
+	}
+
+	if err := c.updateIPEntry(qname, ipv4, lookupTime.Add(time.Duration(minIPv4TTL)), nftables.TypeIPAddr); err != nil {
+		return fmt.Errorf("failed to update IPv4 addresses: %w", err)
+	}
+	if err := c.updateIPEntry(qname, ipv6, lookupTime.Add(time.Duration(minIPv6TTL)), nftables.TypeIP6Addr); err != nil {
+		return fmt.Errorf("failed to update IPv6 addresses: %w", err)
+	}
+
+	return nil
+}
+
+func (c *DNSCache) updateIPEntry(qname string, ips []net.IP, expirationTime time.Time, dtype nftables.SetDatatype) error {
+	var (
+		setName string
+		err     error
+	)
+
+	c.Lock()
+	defer c.Unlock()
+
+	entry, exists := c.fqdnToEntry[qname]
+	if !exists {
+		entry = cacheEntry{}
+	}
+
+	var ipe *ipEntry
+	switch dtype {
+	case nftables.TypeIPAddr:
+		if entry.ipv4 == nil {
+			setName = c.createSetName(qname, dtype.Name, 0)
+			if ipe, err = newIPEntry(setName, expirationTime, dtype); err != nil {
+				return fmt.Errorf("failed to create IPv4 entry")
+			}
+			entry.ipv4 = ipe
+		}
+		ipe = entry.ipv4
+	case nftables.TypeIP6Addr:
+		if entry.ipv6 == nil {
+			setName = c.createSetName(qname, dtype.Name, 0)
+			if ipe, err = newIPEntry(setName, expirationTime, dtype); err != nil {
+				return fmt.Errorf("failed to create IPv6 entry")
+			}
+			entry.ipv6 = ipe
+		}
+		ipe = entry.ipv6
+	}
+
+	if err = ipe.update(setName, ips, expirationTime, dtype); err != nil {
+		return fmt.Errorf("failed to update ipEntry: %w", err)
+	}
+	c.fqdnToEntry[qname] = entry
+
+	return nil
+}
+
+func (c *DNSCache) createSetName(qname, dataType string, suffix int) (setName string) {
+	md5Hash := md5.Sum([]byte(qname + dataType))
+	hex := hex.EncodeToString(md5Hash[:])
+
+	for i, ch := range hex {
+		if !unicode.IsDigit(ch) {
+			setName = hex[i : i+16]
+			break
+		}
+	}
+
+	// Check that set name isn't taken already
+	if _, ok := c.setNames[setName]; ok {
+		setName = c.createSetName(qname, dataType, suffix+1)
+		return
+	}
+
+	c.setNames[setName] = struct{}{}
+	return
+}
+
+// createNftSet creates new Nftables set
+func createNftSet(setName string, dataType nftables.SetDatatype) (err error) {
```

Comment:
is  `createNftSet` really needed? Because sets are completely rendered to the nftables template.